app/controllers/api/submissions/test_runs_controller.rb

```diff
@@ -0,0 +1,23 @@
+module API
+  module Submissions
+    class TestRunsController < BaseController
+      def show
+        submission = current_user.submissions.find_by!(uuid: params[:submission_id])
+
+        test_run = if submission.tests_queued?
```

Comment:
@iHiD Is it fine to place this here? Or should we add it to the serializer?

Reply:
I don't think we should mock out a fake test-run. I think if the test run doesn't exist, we should be returning a `404` with `json: {test_run: nil}` and then the React should deal with that.

I presume the React can already cope with that anyway, as it's the same as the initial state of the exercise?

Reply:
Cool. I will do this in React then. I was just thinking about having some kind of consistency with a "null object" which the server supplies, instead of the client just guessing what it is. It's fine though.

Reply:
I _agree_ with `404`, but i'd return:

```ruby
render json: { errors: [ { message: 'Test run doesn\'t exist.' } ] }, 
           status: :not_found, 
           content_type: 'application/vnd.exercism.errors.v1+json'
```

On the _react_ side, this is a regular 404, with a *structured* error message. You can "just" use `response.json()`, like regular, but you know for a fact that it will be `errors: [ array of errors ]`.

It's probably a _good_ idea, to use `rescue_from ActiveRecord::RecordNotFound` in the `api` base controller, and rescue from with that response. You *can* use a simple gem [`erratum_responsum`](https://github.com/XPBytes/erratum_responsum), which adds all the regular api error codes as messages. This is what we use:

```ruby
  # initializers/<...>.rb  
  self.error_media_type = 'application/vnd.xpbytes.errors.v1+json'

  # api_controller.rb
  include ErratumResponsum

  rescue_from CanCan::AccessDenied, AuthorizedTransaction::TransactionUnauthorized, with: :forbidden
  rescue_from OptimisticallyStale::MissingLockVersion, ActionController::BadRequest,
              RequestMissingParam, with: :bad_request
  rescue_from ActiveRecord::RecordNotFound, with: :not_found
  rescue_from ActionController::UnknownFormat, NoAcceptSerializer, with: :not_acceptable
  rescue_from ResourceGone, with: :gone
  rescue_from ActiveRecord::StatementInvalid, ActiveRecord::RecordInvalid,
              ContentDoesNotMatchContentType, with: :unprocessable_entity
  rescue_from ActiveRecord::RecordNotUnique, ActiveRecord::StaleObjectError, with: :conflict
  rescue_from ContentTypeNotRecognised, with: :unsupported_media_type
```

Additionally, it gives you [_stable_ error codes](https://github.com/XPBytes/erratum_responsum/blob/de6590825bb74625058460819c3944dc4f2cb212/lib/erratum_responsum.rb#L82-L91) you can track:
```ruby
  # inside the gem
  
  def serialize_errors(errors)
    Array(errors).map do |error|
      { message: error.respond_to?(:message) ? error.message : error, code: error_code(error) }
    end
  end

  def error_code(error)
    return 'Ex' + String(error.send(:error_code)) if error.respond_to?(:error_code)
    'Gx' + Digest::MD5.hexdigest(error.class.name)
  end
```

Reply:
This is nice! I'll put this in a separate issue.